packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -71,104 +71,247 @@ export function matches(element: Element, selector: AST): boolean {
+  let ancestor: Element | null = element.parentElement;
+  while (ancestor !== null) {
+    if (ancestor.matches && ancestor.matches(selector)) {
+      return ancestor;
     }
-  } else if (selector.type === 'compound') {
-    // TODO - handling compound needs to be reworked...
-    // .cls:upward(1) for example will not work with this implementation.
-    // :upward is not about selecting, but transforming a set of nodes (i.e.
-    // uBO's transpose method).
-    if (selector.compound.length !== 0) {
-      elements.push(
-        ...querySelectorAll(element, selector.compound[0]).filter((e) =>
-          selector.compound.slice(1).every((s) => matches(e, s)),
-        ),
-      );
+    ancestor = ancestor.parentElement;
+  }
+  return null;
+}
+
+function handleCompoundSelector(element: Element, compound: AST[]): Element[] {
+  if (compound.length === 0) {
+    return [];
+  }
+
+  const upwardIndex = compound.findIndex((s) => s.type === 'pseudo-class' && s.name === 'upward');
+
+  if (upwardIndex === -1) {
+    const [firstSelector, ...restSelectors] = compound;
+    return querySelectorAll(element, firstSelector).filter((e) =>
+      restSelectors.every((s) => matches(e, s)),
+    );
+  }
+
+  const before = compound.slice(0, upwardIndex);
+  const upward = compound[upwardIndex];
+  const after = compound.slice(upwardIndex + 1);
+
+  const candidates =
+    before.length > 0
+      ? querySelectorAll(element, { type: 'compound', compound: before })
+      : [element];
+
+  const ancestors: Element[] = [];
+
+  for (const c of candidates) {
+    if (upward.type !== 'pseudo-class' || upward.name !== 'upward') {
+      continue;
     }
-  } else if (selector.type === 'complex') {
-    const elements2 =
-      selector.left === undefined ? [element] : querySelectorAll(element, selector.left);
-
-    if (selector.combinator === ' ') {
-      for (const element2 of elements2) {
-        elements.push(...querySelectorAll(element2, selector.right));
-      }
-    } else if (selector.combinator === '>') {
-      for (const element2 of elements2) {
-        for (const child of element2.children) {
-          if (matches(child, selector.right) === true) {
-            elements.push(child);
-          }
-        }
+    const { argument } = upward;
+    if (argument === undefined) {
+      continue;
+    }
+
+    const n = Number(argument);
+    const ancestor = !Number.isNaN(n)
+      ? findAncestorByDistance(c, n)
+      : findAncestorBySelector(c, argument);
+
+    if (ancestor === null) {
+      continue;
+    }
+    ancestors.push(ancestor);
+  }
+
+  if (ancestors.length === 0) {
+    return [];
+  }
+
+  if (after.length > 0) {
+    if (after[0].type === 'pseudo-class' && after[0].name === 'upward') {
+      return ancestors.flatMap((a) => querySelectorAll(a, { type: 'compound', compound: after }));
+    }
+    return ancestors.filter((a) => after.every((s) => matches(a, s)));
+  }
```

Comment:
What happens if we have like 4 `:upward` selectors in a row with a mix of multiple other types of selectors? This function picks up the very early `:upward` and handles it. I believe refactoring is required on this function for better readability and clear code flow. Handling upward sections in `after` differently doesn't make sense, despite it seems like it will fall in same function: `handleCompoundSelector` by `querySelectorAll`. Also, then `before` goes empty from second cycle. It seems like there are too much of overhead in this function.